### DISCUSSION TITLE
What is the reason for choosing Rust (altium-rs) over Python (PyAltium)?

### DISCUSSION THREAD

**Discussion by Argmaster** [pluots/altium#15] [Q&A] [answered]
Just asking, no taunt intended.

**Comment by tgross35** [comment, marked as answer]
Totally reasonable question!

I just kind of hit a point in the Python parsing where it was difficult to work with the binary data. I could extract the files alright and do the basics, but it was just a bit hard to get the nuances of binary data when e.g. your integers aren’t sized.

Error handling too, it’s very nice with Rust’s `Result` that you know exactly what kind of errors you might get and have to either handle or propagate them: easier to miss errors that get thrown in Python. Plus I have wrote `.context` that let me add a comment to a “actions backtrace” before propagating it, in one line ([example](https://github.com/pluots/altium/blob/55a969f081a0353bcf9be8c006721e63eb5e88c8/altium/src/sch/storage.rs#L92)]), which is quite helpful.

I plan to expose Python bindings to this library via PyO3 since I think that’s how a lot of people will use it, but rust is just a _really_ good tool for parsing and for repetitive code.

Do you have any use case or are looking to help out? I am very curious how people are using this library

**Reply by Argmaster** [reply]
Unfortunately I don't have any use case (I have nothing to do with PCB design), for now I was just curious, and Your response was definitely convincing. When it comes to Altium format, I was unable to find any specification for it which is a bit discouraging. At the same time I have a lot of work around implementing Gerber format, so I may not have time to get my hands on another project. I'm thankful for Your response and I wish you good luck with the implementation!

**Reply by tgross35** [reply]
It is certainly not very well documented, but https://github.com/vadmium/python-altium/blob/9e3cf5a16150e74ba2b7e6d1f2895e0ce16b9a05/format.md has helped a lot. It is pretty out of date so things like pins are now stored differently, but it at least gives a good idea of what the field names are.

Happy to help!
